# syntax-transform/3.1/0009-separation-of-positional-and-keyword-arguments.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 2, there are three coercion rules for converting between positional
# arguments and keyword arguments that don't exist anymore in Ruby 3.
#
# If a method or block that takes keyword arguments is called without keyword
# arguments, with greater than the minimum required number of positional
# arguments and the last positional argument is a hash that contains no
# non-symbol keys, the last positional argument is converted into keyword
# arguments. For example, this code snippet prints `1` and `{:b => 2}` in Ruby 2
# but throws an `ArgumentError` in later versions.
#
#   def f(a, **kwargs)
#     p a
#     p kwargs
#   end
#   f(1, {:b => 2})
#
# If a method or block that takes keyword arguments is called with keyword
# arguments and fewer than the minimum required number of positional arguments,
# the keyword arguments are collected into a hash and passed as an additional
# keyword argument instead of actually being passed as keyword arguments. For
# example, this code snippet prints `{1=>2}` and `{}` in Ruby 2 but throws an
# `ArgumentError` in later versions.
#
#   def f(a, **kwargs)
#     p a
#     p kwargs
#   end
#   f(1 => 2)
#
# If a method or block that takes keyword arguments and does not have a
# keyword argument splat (e.g. `**kwargs`) is called with keyword arguments,
# with greater than or equal to the minimum required number of positional
# arguments and at least one of the keyword arguments has a non-symbol key, the
# keyword arguments with non-symbol keys are collected into a hash, the hash is
# passed as an additional positional argument and only the keyword arguments
# with symbol keys are actually passed as keyword arguments. For example, this
# code snippet prints `[{3=>4}]` and `2` in Ruby 2 but throws an `ArgumentError`
# in later Ruby versions.
#
#   def f(*args, a: 1)
#     p args
#     p a
#   end
#   f(:a => 2, 3 => 4)

--- a/vm_args.c
+++ b/vm_args.c
@@ -614,6 +614,82 @@ setup_parameters_complex(rb_execution_context_t * const ec, const rb_iseq_t * co
 	break;
     }
 
+    /* If syntax transform is active and targeting Ruby <= 2.7 and a method or block that takes keyword arguments is called */
+    if (mkxp_iseq_is_syntax_transform_active(iseq, 2, 7, -1) && (iseq->body->param.flags.has_kw || iseq->body->param.flags.has_kwrest)) {
+        if (args->kw_argv != NULL) {
+            const VALUE *kw = args->kw_arg->keywords;
+            int kw_len = args->kw_arg->keyword_len;
+            int i;
+            keyword_hash = rb_hash_new_with_size(kw_len);
+            for (i = 0; i < kw_len; ++i) {
+                rb_hash_aset(keyword_hash, kw[i], args->kw_argv[i]);
+            }
+            args->kw_argv = NULL;
+        }
+        /* If the method or block is not called with keyword arguments, is called with more than the minimum required number of positional arguments and the last positional argument is a hash with no non-symbol keys, pass the contents of that hash as the keyword arguments and remove it from the list of positional arguments */
+        if (kw_flag == 0 && given_argc > min_argc && given_argc > 0 && RB_TYPE_P(args->argv[given_argc - 1], T_HASH)) {
+            int kw_len = rb_long2int(RHASH_SIZE(args->argv[given_argc - 1]));
+            struct fill_values_arg arg;
+            arg.keys = ALLOCA_N(VALUE, kw_len * 2);
+            arg.vals = arg.keys + kw_len;
+            arg.argc = 0;
+            rb_hash_foreach(args->argv[given_argc - 1], fill_keys_values, (VALUE)&arg);
+            {
+                int i;
+                for (i = 0; i < kw_len; ++i) {
+                    if (!RB_TYPE_P(arg.keys[i], T_SYMBOL)) {
+                        break;
+                    }
+                }
+                if (i == kw_len) {
+                    rb_warn("Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call");
+                    keyword_hash = args->argv[--given_argc];
+                    --args->argc;
+                }
+            }
+        }
+        /* If the method or block is called with keyword arguments and is called with fewer than the minimum required number of positional arguments, collect the keyword arguments into a hash, append the hash to the list of positional arguments and don't pass any keyword arguments */
+        else if (kw_flag != 0 && given_argc < min_argc) {
+            rb_warn("Passing the keyword argument as the last hash parameter is deprecated");
+            args->argv[given_argc++] = keyword_hash;
+            ++args->argc;
+            keyword_hash = Qnil;
+        }
+        /* If the method or block does not accept keyword rest arguments, is called with keyword arguments, is called with at least the minimum required number of positional arguments and at least one of the passed keyword arguments has a non-symbol key, collect the keyword arguments with non-symbol keys into a hash and append the hash to the list of positional arguments */
+        else if (!iseq->body->param.flags.has_kwrest && kw_flag != 0 && given_argc >= min_argc) {
+            int kw_len = rb_long2int(RHASH_SIZE(keyword_hash));
+            struct fill_values_arg arg;
+            arg.keys = ALLOCA_N(VALUE, kw_len * 2);
+            arg.vals = arg.keys + kw_len;
+            arg.argc = 0;
+            rb_hash_foreach(keyword_hash, fill_keys_values, (VALUE)&arg);
+            {
+                VALUE hash = Qundef;
+                int i;
+                for (i = 0; i < kw_len; ++i) {
+                    if (!RB_TYPE_P(arg.keys[i], T_SYMBOL)) {
+                        rb_hash_delete(keyword_hash, arg.keys[i]);
+                        if (hash == Qundef) {
+                            hash = rb_hash_new_with_size(kw_len);
+                            args->argv[given_argc++] = hash;
+                            ++args->argc;
+                        }
+                        rb_hash_aset(hash, arg.keys[i], arg.vals[i]);
+                    }
+                }
+                if (hash != Qundef) {
+                    if (rb_long2int(RHASH_SIZE(keyword_hash)) == 0) {
+                        rb_warn("Passing the keyword argument as the last hash parameter is deprecated");
+                        keyword_hash = Qnil;
+                    }
+                    else {
+                        rb_warn("Splitting the last argument into positional and keyword parameters is deprecated");
+                    }
+                }
+            }
+        }
+    }
+
     /* argc check */
     if (given_argc < min_argc) {
         if (arg_setup_type == arg_setup_block) {
